### PR TITLE
Refactor Elasticsearch integration

### PR DESCRIPTION
**Pull request by arnisoph**
- [x] check for new incompatibilities
- [x] do final (manual) testing
- [x] test ES connection config
- [x] check documentation

Related issues:
- #20826
- #23125
- #22004
- #21881

### CODE REVIEW

**Comment by bechtoldt** [comment]
Go Go Jenkins!

**Comment by bechtoldt** [comment]
sth. is wrong here

**Comment by rallytime** [comment]
@bechtoldt Are you still working on this one? I am just making the rounds again and it looks like you've made a few more changes.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by bechtoldt** [comment]
@rallytime I was facing #24440 when trying to test my changes. I'll try a newer pyzmq version this week.

**Comment by rallytime** [comment]
Ah yes, I remember commenting on that issue now. As you pull request stands right now, the tests failures are unrelated, so I've added the Tests Passed label. If your testing later this week works out ok without other changes, then this should be good for review/merge.

**Comment by jfindlay** [comment]
@bechtoldt, is this pull request ready to go?

**Comment by bechtoldt** [comment]
@jfindlay please see the checklist in the PR description.

**Comment by bechtoldt** [comment]
At the moment I can't continue working on this PR due to #24440 .

**Comment by bechtoldt** [comment]
I'm open for feedback.

**Comment by thatch45** [comment]
This is awesome @bechtoldt ! I am excited to see that it is ready!
@jfindlay can you please review this?

File: salt/modules/elasticsearch.py [hunk 1]
```diff
@@ -31,77 +65,402 @@ def __virtual__():
...
+
+    if not hosts:
+        hosts = ['127.0.0.1:9200']
+    if isinstance(hosts, string_types):
+        hosts = [hosts]
+    try:
+        es = elasticsearch.Elasticsearch(hosts)
+        if not es.ping():
+            raise CommandExecutionError('Could not connect to Elasticsearch host/ cluster {0}, is it unhealthy?'.format(hosts))
+        log.warn(es.ping())
```
Comment (bechtoldt) [inline salt/modules/elasticsearch.py]: ups, this needs to be removed

**Comment by cachedout** [comment]
Looks outstanding! Merged.

**Comment by bechtoldt** [comment]
Great! Can be backport it to 2015.8?

**Comment by rallytime** [comment]
Sure thing! Done in the PR above.

**Comment by DanyC97** [comment]
Hi @bechtoldt ,

can you please help me understand the pro/cons of using the returner for sending minion logs to Elasticsearch vs logstash ?

Thanks,
Dani

**Pull request merged**